anonymizer/OpenDiffix.Service/Led.fs

```diff
@@ -0,0 +1,225 @@
+module OpenDiffix.Service.Led
+
+open OpenDiffix.Core
+
+/// Compares 2 rows and returns the index where a single column is not equal.
+/// Returns `None` if rows are equal or are different in more than one column.
+let private findSingleNonMatchingColumn max (row1: Row) (row2: Row) =
+  let rec compare notEqualAt i =
+    if i = max then
+      // End of comparison, return matched index (may be None)
+      notEqualAt
+    else if row1.[i] = row2.[i] then
+      // Columns equal, try next one
+      compare notEqualAt (i + 1)
+    else if Option.isSome notEqualAt then
+      // A non-equal column already exists, abort
+      None
+    else
+      // Mark current index as not equal and keep going to verify no others are
+      compare (Some i) (i + 1)
+
+  compare None 0
+
+let private anonymizingAggregatorIndexes (aggregationContext: AggregationContext) =
+  aggregationContext.Aggregators
+  |> Seq.map fst
+  |> Seq.indexed
+  |> Seq.choose (fun (i, aggSpec) -> if Aggregator.isAnonymizing aggSpec then Some i else None)
+  |> Seq.toArray
+
+// For LED hook we merge only anonymizing aggregators.
+let private mergeGivenAggregatorsInto (targetBucket: Bucket) aggIndexes (sourceBucket: Bucket) =
+  let sourceAggregators = sourceBucket.Aggregators
+  let targetAggregators = targetBucket.Aggregators
+
+  for i in aggIndexes do
+    targetAggregators.[i].Merge(sourceAggregators.[i])
+
+type ColumnState = { Values: Dictionary<Value, int>; mutable TopValues: (Value * int) list }
+
+[<Literal>]
+let TOP_VALUES = 3
+
+let rec private removeValue index value list =
+  if index = TOP_VALUES then
+    [] // Trim off excess
+  else
+    match list with
+    | [] -> []
+    | (topValue, _topCount) :: tail when topValue = value -> tail
+    | head :: tail -> head :: removeValue (index + 1) value tail
+
+let rec private insertValueSorted index (value, count) list =
+  if index = TOP_VALUES then
+    [] // Trim off excess
+  else
+    match list with
+    | [] -> [ value, count ]
+    | (_topValue, topCount) :: _tail when count > topCount -> (value, count) :: removeValue (index + 1) value list
+    | head :: tail -> head :: insertValueSorted (index + 1) (value, count) tail
+
+let private isolatingColumns (aggregationContext: AggregationContext) (buckets: Bucket seq) =
+  let countsByColumn =
+    Array.init aggregationContext.GroupingLabels.Length (fun _ -> { Values = Dictionary(); TopValues = [] })
+
+  let mutable totalRows = 0
+
+  let buckets =
+    buckets
+    |> Utils.safeIter (fun bucket ->
+      let count = bucket.RowCount
```

Comment:
if possible, can we refactor this into a well-named function to document this step? Hm, actually this comment applies to all the lambdas here. If we could extract into named functions, it would be easier to grasp the overall flow of the algo

Reply:
Should we do local functions or named functions with more verbose params?

Reply:
no idea :/, I always ask myself the same question, and I can hardly make up my mind in most cases :joy:. I guess it depends on how many variables from the outside scope are used and what is the overall layout and role of the function.